Mathlib/Topology/Algebra/ProperAction.lean

```diff
@@ -0,0 +1,342 @@
+  isProperMap_vadd_pair' : IsProperMap (fun gx ↦ ⟨gx.1 +ᵥ gx.2, gx.2⟩ : G × X → X × X)
+
+/-- Proper action in the sense of Bourbaki:
+the map `G×X→ X×X` is a proper map `isProperMap`
+-/
+@[to_additive existing, mk_iff]
+class ProperSMul (G X : Type*) [TopologicalSpace G] [TopologicalSpace X] [Group G]
+    [MulAction G X] : Prop where
+  isProperMap_smul_pair' : IsProperMap (fun gx ↦ ⟨gx.1 • gx.2, gx.2⟩ : G × X → X × X)
+
+attribute [to_additive existing] properSMul_iff
+
+/-- By definition, if G acts properly on X
+the map `G×X→ X×X` is a proper map `isProperMap`
+-/
+@[to_additive]
+lemma isProperMap_smul_pair (G X : Type*) [Group G] [MulAction G X]
+    [TopologicalSpace G] [TopologicalSpace X] [ProperSMul G X] :
+    IsProperMap (fun gx ↦ ⟨gx.1 • gx.2, gx.2⟩ : G × X → X × X) :=
+  ProperSMul.isProperMap_smul_pair'
+
+open Filter Topology
+
+variable {G X Y Z W : Type*} [Group G] [MulAction G X] [MulAction G Y]
+variable [TopologicalSpace G] [TopologicalSpace X] [TopologicalSpace Y]
+variable [TopologicalSpace Z] [TopologicalSpace W]
+
+@[to_additive]
+instance continuousSmul_of_properSMul [ProperSMul G X] : ContinuousSMul G X where
+  continuous_smul := (isProperMap_smul_pair G X).continuous.fst
+
+/-- A group `G` acts properly on a topological space `X` if and only if for all ultrafilters`𝒰` on `X × G`, if `𝒰` converges to `(x₁, x₂)` along the map `(g, x) ↦ (g • x, x)`,
+then there exists `g : G` such that `g • x₂ = x₁` and `𝒰.fst` converges to `g`. -/
+@[to_additive "A group acts `G` properly on a topological space `X` if and only if
+for all ultrafilters `𝒰` on `X`, if `𝒰` converges to `(x₁, x₂)`
+along the map `(g, x) ↦ (g • x, x)`, then there exists `g : G` such that `g • x₂ = x₁`
+and `𝒰.fst` converges to `g`."]
+theorem properSMul_iff_continuousSMul_ultrafilter_tendsto : ProperSMul G X ↔ ContinuousSMul G X
+    ∧ (∀ 𝒰 : Ultrafilter (G × X), ∀ x₁ x₂ : X,
+    Tendsto (fun gx ↦ ⟨gx.1 • gx.2, gx.2⟩ : G × X → X × X) 𝒰 (𝓝 (x₁, x₂)) →
+    ∃ g : G, g • x₂ = x₁ ∧ Tendsto Prod.fst (𝒰 : Filter (G × X)) (𝓝 g)) := by
+  constructor
+  · intro h
+    refine ⟨by infer_instance, fun 𝒰 x₁ x₂ h' ↦ ?_⟩
+    rw [properSMul_iff, isProperMap_iff_ultrafilter] at h
+    have ⟨(g, x), hgx1, hgx2⟩ := h.2 h'
+    use g
+    constructor
+    · simp at hgx1
+      rw [← hgx1.2, hgx1.1]
+    · have := continuous_fst.tendsto (g, x)
+      rw [Tendsto] at *
+      calc
+        map Prod.fst ↑𝒰 ≤ map Prod.fst (𝓝 (g, x)) := map_mono hgx2
+        _               ≤ 𝓝 (g, x).1 := this
+  · rintro ⟨cont, h⟩
+    rw [properSMul_iff, isProperMap_iff_ultrafilter]
+    refine ⟨by fun_prop, fun 𝒰 (x₁, x₂) hxx ↦ ?_⟩
+    rcases h 𝒰 x₁ x₂ hxx with ⟨g, hg1, hg2⟩
+    use (g, x₂)
+    refine ⟨by rw [hg1], ?_⟩
+    rw [nhds_prod_eq, 𝒰.le_prod]
+    refine ⟨hg2, ?_⟩
+    change Tendsto (Prod.snd ∘ (fun gx : G × X ↦ (gx.1 • gx.2, gx.2))) ↑𝒰 (𝓝 (Prod.snd (x₁, x₂)))
+    apply Filter.Tendsto.comp
+    apply Continuous.tendsto
+    exact continuous_snd
+    assumption
```

Comment:
```suggestion
    rw [nhds_prod_eq, 𝒰.le_prod]
    exact ⟨hg2, (continuous_snd.tendsto _).comp hxx⟩
```

Notice that the `change` is necessary for the `apply Filter.Tendsto.comp` to work, because Lean essentially has to guess how you want to write your function as a composition. But if you do everything at once, then Lean can just check the type of your term and then check that it is equal to what was expected.

Alternatively, you can also do:
```suggestion
    simp_rw [nhds_prod_eq, Filter.le_prod, Filter.tendsto_prod_iff'] at hxx ⊢
    exact ⟨hg2, hxx.2⟩
```
The reason I use `simp_rw` instead of `rw` is that `rw` expects that each lemma is applicable to *all* places that you're rewriting, while `simp_rw` only wants that each lemma is applicable to *at least* one place.